prototype2/pubsub/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+                                                      -> Publisher<'a, T>;
+  fn read_subscriber<'a, T: Clone + Any + Send + Sync>(&'a mut self,
+                                                       token: &SubscriberToken<T>)
+                                                       -> Subscriber<'a, T>;
+
+  fn register_publisher<T: Clone + Any + Send + Sync>(&mut self) -> PublisherToken<T>;
+  fn register_subscriber<T: Clone + Any + Send + Sync>(&mut self) -> SubscriberToken<T>;
+}
+
+/// An indicator that the holder can access a pubsub resource and retrieve a
+/// provider corresponding
+/// to this type
+pub struct PublisherToken<T> {
+  phantom_data: PhantomData<T>,
+}
+
+/// An indicator pointing to a specific subscriber reader that the holder has
+/// access to.
+pub struct SubscriberToken<T> {
+  phantom_data: PhantomData<T>,
+  reader_idx: u32,
+}
+
+struct PubSubRegistry {
+  readers: HashMap<TypeId, u32>,
+}
+
+
+impl PubSubRegistry {
+  fn new() -> PubSubRegistry {
+    PubSubRegistry { readers: HashMap::new() }
+  }
+}
+
+impl PubSubStore for specs::World {
+  /// Attempt to retrieve the general publisher for this resource
+  fn read_publisher<'a, T: Clone + Any + Send + Sync>(&'a mut self,
+                                                      token: &PublisherToken<T>)
+                                                      -> Publisher<'a, T> {
+
+    Publisher { senders: acquire_raw_containers::<T>(self) }
+  }
+
+  /// Attempt to retrieve the subscriber corresponding to this token
+  fn read_subscriber<'a, T: Clone + Any + Send + Sync>(&'a mut self,
+                                                       token: &SubscriberToken<T>)
+                                                       -> Subscriber<'a, T> {
+
+    Subscriber {
+      receivers: acquire_raw_containers::<T>(self),
+      idx: token.reader_idx,
+    }
+  }
+
+  /// Provide a publisher token for the user. We don't actually record this,
+  /// its merely to make
+  /// sure they follow procedure to get access to the resource.
+  fn register_publisher<T: Clone + Any + Send + Sync>(&mut self) -> PublisherToken<T> {
+    PublisherToken { phantom_data: PhantomData }
+  }
+
+  /// Allocate the subscriber a Receiver and provide them a token containing
+  /// that receiver index.
+  fn register_subscriber<T: Clone + Any + Send + Sync>(&mut self) -> SubscriberToken<T> {
+    let idx = {
+      let mut registry = acquire_registry(self);
+
+      let type_id = TypeId::of::<T>();
+      let mut idx = registry.readers.entry(type_id).or_insert(0);
+      *idx += 1;
+      idx.clone() - 1
+    };
+
+    acquire_raw_containers::<T>(self).push(NewContainer(Vec::new()));
+
+    SubscriberToken {
+      phantom_data: PhantomData,
+      reader_idx: idx,
+    }
+  }
+}
+
+// Wrap builtins so they cannot be accessed outside this module
+// This prevents accidental access by someone who happens to be storing Vecs
+struct NewContainer<T: Clone + Any + Send + Sync>(pub Vec<T>);
```

Comment:
I don't want anyone getting too attached to it LOL. If you've got a better one I'll take it.